Run processing algorithms using task manager

nyalldawson opened the pull request:
And make processing algorithm execution safe to run in background threads.

nyalldawson committed:
Add a pure virtual clone method for algorithms

This is required for safely executing the algorithm in a background
thread.

nyalldawson committed:
Fix re-saving already saved models

nyalldawson committed:
Improvements to QgsProcessingAlgRunnerTask

- take a clone of algs before running them. This avoids issues
if the algorithm is removed or edited while a background task
is running

- accept an optional existing feedback object

nyalldawson committed:
Add some more clone methods to algorithm subclasses

nyalldawson committed:
Fix some processing algorithm exception handling

nyalldawson committed:
Add clone method to ScriptAlgorithm

nyalldawson committed:
Split algorithm execution into separate prepare/process/postProcess steps

The prepare and postProcess steps are designed to be run in main
thread only, while the process step can safely be run in a background
thread.

nyalldawson committed:
Upgrade ported python algs to be thread ready

nyalldawson committed:
Rename QgsProcessingAlgorithm::clone to ::create

Since it better describes what the function does. It returns a new
pristine copy of the algorithm, not a clone of its current state

nyalldawson committed:
Flush sink buffers instead of deleting sink/sources in processAlgorithm

Avoids potential issues if processAlgorithm is run in a different
thread

nyalldawson committed:
Fix tests

nyalldawson committed:
Make algorithm dialog use background tasks to execute algorithms

nyalldawson committed:
Better assert messages

Issues is actually when prepare/postProcess is called in a different
thread to context - it doesn't have to be the main thread

nyalldawson committed:
Add method to transfer all layers from one map store to another

With a note and assert that both stores must have the same
thread affinity

nyalldawson committed:
Add methods to retrieve current thread affinity and push contexts
to another thread

With suitable assert in place to ensure that pushes are only
made when current thread == existing thread affinity

nyalldawson committed:
Add method to copy thread safe settings between processing contexts

nyalldawson committed:
Add method to take results from another processing context and
add to the current context

With appropriate note and tests to ensure that both the current
context and that which the results being taken from share the
same thread affinity

nyalldawson committed:
Remove unused member

nyalldawson committed:
Make processing algorithms safe to run in threads

nyalldawson committed:
Make QgsProcessingAlgRunnerTask work correctly

It now safely can execute algorithms in background threads
without issues

nyalldawson committed:
Don't set thinking cursors when running processing algorithms

Since now they're run in a background thread, it's not appropriate
anymore

nyalldawson committed:
Avoid QgsFeedback flooding progress report signals

We only emit the progress changed signal when there's been at
least a 0.1% change since the last progress report. Otherwise
it's possible that things like processing algorithms which
are reporting feedback every feature in a 500k feature dataset
cause a gazillion signals to be emitted and everything to
slow to a crawl

nyalldawson commented:
@wonder-sk I'd appreciate a review here :)

Pertinent commits are:

- https://github.com/qgis/QGIS/pull/4799/commits/112a7321a6da3b5151ad80de30fe30940764e645
- https://github.com/qgis/QGIS/pull/4799/commits/d642e5d851d72a37486441c9b21171e656addbbf
- https://github.com/qgis/QGIS/pull/4799/commits/66336fb037d5c207f9b99dca83b40830da535020
- https://github.com/qgis/QGIS/pull/4799/commits/78382c45dfceabab5f0adaa66eef27437a9f1195
- https://github.com/qgis/QGIS/pull/4799/commits/bca03a5778b3bd570eaba86ed89c65846936e50d
- https://github.com/qgis/QGIS/pull/4799/commits/77daede39049a15525768a399fbd58121d2ef465

nyalldawson committed:
Fix build

nyalldawson commented:
@wonder-sk I believe there's still one thing we need to address here - while we agreed that adding layers to a store with a different thread affinity is a bad thing, I think we still need to make sure that accessing layers from a store (and hence QgsProject) is safe to do from any thread. It'd be a matter of protecting the QMap member in QgsMapLayerStore with some appropriate read/write mutexes so that there's no possible issue if someone tries to retrieve a layer from a different thread while a layer is being added/ removed to the store from the main thread.

nirvn commented:
OMG, this is the promised land, we're about to reach it! 😉 

This will work so well with your status bar locator feature.

nirvn commented:
@nyalldawson, testing this PR out -- things are looking good. I however noticed that the progress bar for the zonal stats alg isn't working anymore. The whole UI freezes until the process is over.

nyalldawson committed:
Fix execution of script algorithms was not releasing layers/sinks

nyalldawson committed:
Better debuging from processing alg tests

nyalldawson committed:
Fix procesing test layer loading

nyalldawson committed:
Fix zonal stats algorithm execution

nyalldawson committed:
Keep correct order for zonal stats options, set some stats by default

nyalldawson committed:
Indentation

nyalldawson committed:
Fix build warning

luipir commented on src/core/processing/qgsprocessingalgorithm.cpp:
```diff
@@ -361,25 +361,74 @@ bool QgsProcessingAlgorithm::runPrepared( QgsProcessingContext &context, QgsProc
+  QgsProcessingContext *runContext = nullptr;
+  if ( context.thread() == QThread::currentThread() )
+  {
+    // OH. No issues. Seems you're running everything in the same thread, so go about your business. Sorry about
```
In this way it's a pleasure read the code :) 
super +1

nyalldawson committed:
Algorithms don't have to be split to prepare/process/postProcess

Since it's safe to evaluate parameters in background threads
now, it's usually going to be ok to evaluate everything in
the processAlgorithm step.

This keeps the algorithm code as simple as possible, and will
make porting faster.

Note that the prepare/postProcess virtual methods still exist
and can be used when an algorithm MUST do setup/cleanup work
in the main thread.

nyalldawson committed:
Use a feature source instead of vector layer for QgsGeometrySnapper

nyalldawson committed:
Add equality operator for QgsProcessingFeatureSourceDefinition

nyalldawson committed:
Don't try to load default styles in processing test layers

nyalldawson committed:
Add missing signal

nyalldawson committed:
When running algorithm tests, if two parameters share the same
layer source, ensure that the actual parameter values point
to the same layer

nyalldawson committed:
Port snap geometries algorithm to new API

nyalldawson committed:
runPrepared rethrows exceptions

nyalldawson committed:
Port voronoi polygons algorithm to new API

nyalldawson committed:
Port delaunay triangulation alg to new API

nyalldawson committed:
Port a multi-step algorithm to new API (concave hull)

nyalldawson commented:
> , testing this PR out -- things are looking good. I however noticed that the progress bar for the zonal stats alg isn't working anymore. The whole UI freezes until the process is over.

That's fixed now

nyalldawson committed:
Add a method to take result layers (and ownership) from processing context

nyalldawson committed:
Make concave hull alg more efficient

- remove temporary layers from context, delete them as soon as they
are finished with
- directly remove features via data provider, instead of selecting
and using edit buffer
- use native geometry methods for splitting to single features
and removing rings

nyalldawson committed:
When calling procesing.run(), map layer results are automatically
converted to QgsMapLayer objects with the ownership transferred
to the Python caller

This should make it super-easy for PyQGIS scripts to run processing
algorithms and immediately utilise the results, even if they
are memory layers. They call processing.run(), and get a dict of
results back which includes those layers ready for adding to the
current project or doing some other processing or operations
with, and if they don't transfer to ownership off then these
layers will be correctly garbaged collected by Python.

nyalldawson committed:
Fix processing.runAndLoadResults

nyalldawson committed:
Fix possible error when messagebar items are popped and events are queued

nyalldawson committed:
Fix deadlock when running algs in task manager and python exception occurs

gioman commented:
@nyalldawson sorry for the naive question, does this applies only to qgis native algs or also any other tool (even from external providers)?

nyalldawson commented:
>  sorry for the naive question, does this applies only to qgis native algs or also any other tool (even from external providers)?

Not naive at all! It should apply (transparently) to any providers.

nyalldawson commented:
@gioman When this is reviewed and merged I'll pull in the commits from https://github.com/nyalldawson/QGIS/tree/more_processing which bring across 2 gdal algs. That should open the door for the rest to be ported :)

gioman commented:
> Not naive at all! It should apply (transparently) to any providers.

I can't find a good enough emoji to react to this... thanks a thousand tons!

nyalldawson commented:
> I can't find a good enough emoji to react to this... thanks a thousand tons!

Before you get too excited - keep in mind that this doesn't "multithread" models. So a whole model is now run in a background thread, but there's no logic (yet) to run non-dependant parts in parallel. Unfortunately that part is out of scope for my current work, but should be doable later without API break issues.

gioman commented:
> Before you get too excited - keep in mind that this doesn't "multithread" models. So a whole model is now run in a background thread, but there's no logic (yet) to run non-dependant parts in parallel. Unfortunately that part is out of scope for my current work, but should be doable later without API break issues.

I understand, thanks. But (unless I got it very wrong) this anyway means that any single tool/algorithm will run in the background, not blocking qgis main window, correct?

nyalldawson commented:
> I understand, thanks. But (unless I got it very wrong) this anyway means that any single tool/algorithm will run in the background, not blocking qgis main window, correct?

Yes. Well.... when run through the algorithm dialog anyway. There's probably other parts i need to switch away from direct `processing.run(...)` commands. But I'll check that when this PR is merged.

I'd love early tests if you feel like building!

gioman commented:
> Yes. Well.... when run through the algorithm dialog anyway

this is by itself a *huge* step forward.

gioman commented:
> I'd love early tests if you feel like building!

no problem, I will.

wonder-sk commented:
@nyalldawson I have had a look at the commits you highlighted - it all looks great.

I would just suggest to add few more bits to the api docs:
- prepare + post-process - what kind of preparation/post-processing to do (e.g. make feature sources from layers?) - to help alg authors understand what to put into those implementations and what not
- is "prepare" allowed to fail? if it fails, what happens next?
- "post-process" - is it called if alg failed? (from what I understand, no)

For your other comment to allow access to layer stores from a different thread, I probably do not understand why do we need it - isn't it enough to just push such layer store to the thread where it will be needed? I think it is still a bad idea in any thread to do anything with QObject subclasses that belong to another thread (or we risk that bad things happen due to unexpected signal/slot behavior).

nyalldawson committed:
Expand on docs

nyalldawson committed:
Update tests
